lib/miq_automation_engine/models/miq_ae_browser.rb

```diff
@@ -0,0 +1,139 @@
+class MiqAeBrowser
+  attr_accessor :user
+
+  # Automate Datastore Browser starting at MiqAeDomain, down to the MiqAeInstances and MiqAeMethods.
+
+  # Searches directory ala LDAP style
+  #   Starting point of searches using fqname  (dn)
+  #   Supporing scope base, one and sub
+  # Returns array of objects
+  #
+  # Supporting options:
+  #   :serialize
+  #   :state_machines
+  #
+  # :serialize returns the Hash serialized objects including fqname, domain_fqname and klass
+  # i.e.
+  #   {
+  #     "fqname": "/ManageIQ/System/Request",
+  #     "domain_fqname": "/System/Request",       # fqname relative to /domain
+  #     "klass": "MiqAeClass",
+  #     "id": 75,
+  #     "description": "Automation Requests",
+  #     "name": "Request",
+  #     "created_on": "2015-12-09T20:56:44Z",
+  #     "updated_on": "2015-12-09T20:56:44Z",
+  #     "namespace_id": 41,
+  #     "updated_by": "system"
+  #   }
+  #
+  # with :state_machines true, only the objects belonging to a sub-tree that
+  # includes state machine entrypoints are returned.
+
+  SCOPE_BASE = "base".freeze
+  SCOPE_ONE  = "one".freeze
+  SCOPE_SUB  = "sub".freeze
+  SCOPES = [SCOPE_BASE, SCOPE_ONE, SCOPE_SUB].freeze
+
+  def initialize(user = User.current_user)
+    raise "Must be authenticated before using #{self.class.name}" unless user
+    @user = user
+    @waypoint_ids = MiqAeClass.waypoint_ids_for_state_machines
+  end
+
+  def domains(options = {})
+    filter_ae_objects(@user.current_tenant.visible_domains, options)
+  end
+
+  def search(object_ref = nil, options = {})
+    validate_options(options)
+    object = object_ref
+    if object_ref.kind_of?(String)
+      if object_ref.blank? || object_ref == "/"
+        object = nil
+      else
+        object = find_base_object(object_ref, options)
+        raise "Invalid Automate object path #{object_ref} specified to search" if object.blank?
+      end
+    end
+    options[:serialize] ? _search_serialize(object, options) : _search(object, options)
+  end
+
+  private
+
+  def _search_serialize(object = nil, options = {})
```

Comment:
Avoid using `_` as it has special meaning in Ruby... No need to make private methods start with `_` anyway.